src/com/ms/silverking/cloud/dht/net/ProtoRetrievalMessageGroup.java

```diff
@@ -77,93 +89,128 @@ public ProtoRetrievalMessageGroup(UUIDBase uuid, long context, RetrievalOptions
+  }
+
+  public static int getSTLength(MessageGroup mg) {
+    return getOptionBuffer(mg).getShort(RetrievalMessageFormat.stDataOffset);
+  }
+
+  private static Set<SecondaryTarget> getSecondaryTargets(MessageGroup mg) {
+    int stLength;
+    byte[] stDef;
 
-    public static int getUOLength(MessageGroup mg) {
-        int start = getUOStart(mg);
-        return mg.getBuffers()[optionBufferIndex].getInt(start);
+    stLength = getSTLength(mg);
+    if (stLength == 0) {
+      return DHTConstants.noSecondaryTargets;
+    } else {
+      stDef = new byte[stLength];
+      System.arraycopy(getOptionBuffer(mg).array(), RetrievalMessageFormat.stDataOffset + NumConversion.BYTES_PER_SHORT,
+          stDef, 0, stLength);
+      return SecondaryTargetSerializer.deserialize(stDef);
     }
+  }
 
-    private static byte[] getUserOptions(MessageGroup mg) {
-        int uoLength;
-        byte[] uo;
-        int uoStartPos;
-
-        uoStartPos = getUOStart(mg);
-        uoLength   = getUOLength(mg);
-        if (uoLength == 0) {
-            return RetrievalOptions.noUserOptions;
-        } else {
-            uo = new byte[uoLength];
-            System.arraycopy(mg.getBuffers()[optionBufferIndex].array(),
-                uoStartPos + NumConversion.BYTES_PER_INT, uo, 0, uoLength);
-            return uo;
-        }
+  private static int getUOStart(MessageGroup mg) {
+    return RetrievalMessageFormat.stDataOffset + NumConversion.BYTES_PER_SHORT + getSTLength(mg);
+  }
+
+  private static int getAUStart(MessageGroup mg){
+    return getUOStart(mg) + NumConversion.BYTES_PER_INT + getUOLength(mg);
+  }
+
+  private static int getUOLength(MessageGroup mg) {
+    int start = getUOStart(mg);
+    return getOptionBuffer(mg).getInt(start);
+  }
+
+  private static int getAULength(MessageGroup mg){
+    int start = getAUStart(mg);
+    return getOptionBuffer(mg).getInt(start);
+  }
+
+  private static byte[] getUserOptions(MessageGroup mg) {
+    int uoLength;
+    byte[] uo;
+    int uoStartPos;
+
+    uoStartPos = getUOStart(mg);
+    uoLength = getUOLength(mg);
+    if (uoLength == 0) {
+      return RetrievalOptions.noUserOptions;
+    } else {
+      uo = new byte[uoLength];
+      ByteBuffer opts = getOptionBuffer(mg);
+      opts.position(uoStartPos+NumConversion.BYTES_PER_INT);
+      opts.get(uo, 0, uoLength);
+      return uo;
     }
+  }
+
+  private static byte[] getAuthorizationUser(MessageGroup mg) {
+    byte[] au;
+    int auLength = getAULength(mg);
+    int auStartPos = getAUStart(mg);
 
-    protected MessageGroup toMessageGroup(boolean flip) {
-        MessageGroup    mg;
-        
-        mg = super.toMessageGroup(flip);
-        return mg;
+    if (auLength == 0){
+      return RetrievalOptions.noAuthorizationUser;
+    } else {
+      au = new byte[auLength];
+      ByteBuffer opts = getOptionBuffer(mg);
+      opts.position(auStartPos+NumConversion.BYTES_PER_INT);
+      opts.get(au, 0, auLength);
+      return au;
     }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 199:207, 183:191.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.